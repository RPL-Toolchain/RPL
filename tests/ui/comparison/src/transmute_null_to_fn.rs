//@revisions: inline normal
//@[inline] compile-flags: -Zinline-mir=false
#[cfg_attr(test, test)]
fn base_case() {
    unsafe {
        let x: fn() = std::mem::transmute(0 as *const u64);
        //~^ transmute_null_to_fn
        dbg!(x);

        let x: fn() = std::mem::transmute(std::ptr::null::<u64>());
        //~^ transmute_null_to_fn
        dbg!(x);
    }
}

#[cfg_attr(test, test)]
fn cross_function_null_ptr() {
    const fn const_null_ptr<T>() -> *const T {
        std::ptr::null()
    }
    fn null_ptr<T>() -> *const T {
        std::ptr::null()
    }
    unsafe {
        let x: fn() = std::mem::transmute(const_null_ptr::<u64>());
        //~^ transmute_null_to_fn
        dbg!(x);

        let x: fn() = std::mem::transmute(null_ptr::<u64>());
        //~^ transmute_null_to_fn
        dbg!(x);
    }
}

#[cfg_attr(test, test)]
fn cross_statement() {
    let null_ptr = 0 as *const u64;
    unsafe {
        let x: fn() = std::mem::transmute(null_ptr);
        //~^ transmute_null_to_fn
        dbg!(x);
    }

    let null_ptr = std::ptr::null::<u64>();
    unsafe {
        let x: fn() = std::mem::transmute(null_ptr);
        //~^ transmute_null_to_fn
        dbg!(x);
    }
}

pub(crate) fn main() {
    base_case();
    cross_function_null_ptr();
    cross_statement();
}
