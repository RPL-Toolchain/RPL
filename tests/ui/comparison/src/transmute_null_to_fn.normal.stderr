error: transmuting a known null pointer into a function pointer
  --> tests/ui/comparison/src/transmute_null_to_fn.rs:6:23
   |
LL |         let x: fn() = std::mem::transmute(0 as *const u64);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `0 as *const u64` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior
   = note: `-D rpl::transmute-null-to-fn` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rpl::transmute_null_to_fn)]`

error: transmuting a known null pointer into a function pointer
  --> tests/ui/comparison/src/transmute_null_to_fn.rs:10:23
   |
LL |         let x: fn() = std::mem::transmute(std::ptr::null::<u64>());
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `std::ptr::null::<u64>()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/comparison/src/transmute_null_to_fn.rs:39:23
   |
LL |         let x: fn() = std::mem::transmute(null_ptr);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `0 as *const u64` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: transmuting a known null pointer into a function pointer
  --> tests/ui/comparison/src/transmute_null_to_fn.rs:46:23
   |
LL |         let x: fn() = std::mem::transmute(null_ptr);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: try wrapping your function pointer type in `Option<T>` instead, and using `None` as a null pointer value
   = note: `std::ptr::null::<u64>()` is known to be a null pointer, so transmuting it into a function pointer is undefined behavior

error: aborting due to 4 previous errors

